x-pack/plugin/logsdb/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/DataStreamLicenseUpgradeIT.java

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.logsdb;
+
+import org.elasticsearch.index.mapper.SourceFieldMapper;
+
+import java.io.IOException;
+import java.util.List;
+
+public class DataStreamLicenseUpgradeIT extends DataStreamLicenseChangeIT {
+    @Override
+    protected void applyInitialLicense() {}
+
+    @Override
+    protected void licenseChange() throws IOException {
+        startTrial();
+    }
+
+    @Override
+    protected List<TestCase> cases() {
+        return List.of(new TestCase() {
```

Comment:
Some of the test cases are shared with downgrade tests, with the outcome flipped or adjusted. Maybe they can be futher refactored, with upgrade/downgrade just giving a different pair of initial/final mode.